File: Java-Core/src/main/java/presto/statement/WriteStatement.java
package presto.statement;

import presto.error.InvalidResourceError;
import presto.error.NullReferenceError;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.type.IType;
import presto.type.ResourceType;
import presto.type.VoidType;
import presto.utils.CodeWriter;
import presto.value.IResource;
import presto.value.IValue;

public class WriteStatement extends SimpleStatement {

	IExpression content;
	IExpression resource;
	
	public WriteStatement(IExpression content, IExpression resource) {
		this.content = content;
		this.resource = resource;
	}
	

	@Override
	public void toDialect(CodeWriter writer) {
		writer.append("write ");
		switch(writer.getDialect()) {
		case E:
		case S:
			content.toDialect(writer);
			break;
		case O:
			writer.append("(");
			content.toDialect(writer);
			writer.append(")");
			break;
		}
		writer.append(" to ");
		resource.toDialect(writer);
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		context = context.newResourceContext();
		IType resourceType = resource.check(context);
		if(!(resourceType instanceof ResourceType))
			throw new SyntaxError("Not a resource!");
		return VoidType.instance();
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		context = context.newResourceContext();
		Object o = resource.interpret(context);
		if(o==null)
			throw new NullReferenceError();
		if(!(o instanceof IResource))
			throw new InternalError("Illegal write source: " + o);
		IResource res = (IResource)o;
		if(!res.isWritable())
			throw new InvalidResourceError("Not writable");
		res.writeFully(content.interpret(context).toString());
		return null;
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/statement/WriteStatement.js
var SimpleStatement = require("./SimpleStatement").SimpleStatement;
var ResourceType = require("../type/ResourceType").ResourceType;
var VoidType = require("../type/VoidType").VoidType;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var NullReferenceError = require("../error/NullReferenceError").NullReferenceError;
var InternalError = require("../error/InternalError").InternalError;
var InvalidResourceError = require("../error/InvalidResourceError").InvalidResourceError;

function WriteStatement(content, resource) {
	SimpleStatement.call(this);
	this.content = content;
	this.resource = resource;
	return this;
}

WriteStatement.prototype = Object.create(SimpleStatement.prototype);
WriteStatement.prototype.constructor = WriteStatement;

WriteStatement.prototype.toString = function() {
	return "write " + this.content.toString() + " to " + this.resource.toString();
};

WriteStatement.prototype.check = function(context) {
	context = context.newResourceContext();
	var resourceType = this.resource.check(context);
	if(!(resourceType instanceof ResourceType))
		throw new SyntaxError("Not a resource!");
	return VoidType.instance;
}

WriteStatement.prototype.interpret = function(context) {
	context = context.newResourceContext();
	var o = this.resource.interpret(context);
	if(o==null) {
		throw new NullReferenceError();
	}
	if(!(o.isWritable)) {
		throw new InternalError("Illegal write source: " + o);
	}
	if(!o.isWritable()) {
		throw new InvalidResourceError("Not writable");
	}
	var str = this.content.interpret(context).toString();
	o.writeFully(str);
	return null;
};

WriteStatement.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

WriteStatement.prototype.toEDialect = function(writer) {
    writer.append("write ");
    this.content.toDialect(writer);
    writer.append(" to ");
    this.resource.toDialect(writer);
};

WriteStatement.prototype.toODialect = function(writer) {
    writer.append("write (");
    this.content.toDialect(writer);
    writer.append(") to ");
    this.resource.toDialect(writer);
};

WriteStatement.prototype.toSDialect = function(writer) {
    this.toEDialect(writer);
};

exports.WriteStatement = WriteStatement;
